src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs

```diff
@@ -63,5 +78,39 @@ internal override async Task<IPackageSearchMetadata> GetPackageMetadataAsync<T>(
 
             return await base.GetPackageMetadataAsync(identity, includePrerelease, cancellationToken);
         }
+
+        internal static PackageCollectionItem GetLatestApplicableTransitivePackageVersion(IGrouping<string, PackageCollectionItem> groupedPackagesById, IEnumerable<PackageCollectionItem> installedPkgs)
+        {
+            if (groupedPackagesById == null || !groupedPackagesById.Any())
+            {
+                throw new ArgumentException(Strings.Argument_Cannot_Be_Null_Or_Empty, nameof(groupedPackagesById));
+            }
+            if (installedPkgs == null)
+            {
+                throw new ArgumentNullException(nameof(installedPkgs));
+            }
+            if (groupedPackagesById
+                    .Any(group => !group.PackageReferences.Any() // PackageReferences is empty
+                        || group.PackageReferences.Any(pkgRef =>
+                            pkgRef is not ITransitivePackageReferenceContextInfo tr // it is not a transitive package
+                            || !tr.TransitiveOrigins.Any()))) // No transitive origin for a transitive package
```

Comment:
This is going to check all of the PackageReferences to see if they're ITransitivePackageReferenceContextInfo. Then later, we do the same thing but actually cast them to ITransitivePackageReferenceContextInfo. Can we just handle this all in one iteration and fail out if we encounter anything that isn't castable to ITransitivePackageReferenceContextInfo to avoid iterating over it twice and having duplicate Linq allocations?